server/src/main/java/org/elasticsearch/cluster/coordination/ElectionSchedulerFactory.java

```diff
@@ -0,0 +1,203 @@
+    private static final String ELECTION_MAX_TIMEOUT_SETTING_KEY = "cluster.election.max_timeout";
+
+    /*
+     * The first election is scheduled to occur a random number of milliseconds after the scheduler is started, where the random number of
+     * milliseconds is chosen uniformly from
+     *
+     *     (0, min(ELECTION_INITIAL_TIMEOUT_SETTING, ELECTION_MAX_TIMEOUT_SETTING_KEY)]
+     *
+     * For `n > 1`, the `n`th election is scheduled to occur a random number of milliseconds after the `n - 1`th election, where the random
+     * number of milliseconds is chosen uniformly from
+     *
+     *     (0, min(ELECTION_INITIAL_TIMEOUT_SETTING + (n-1) * ELECTION_BACK_OFF_TIME_SETTING, ELECTION_MAX_TIMEOUT_SETTING_KEY)]
+     */
+
+    public static final Setting<TimeValue> ELECTION_INITIAL_TIMEOUT_SETTING = Setting.timeSetting(ELECTION_INITIAL_TIMEOUT_SETTING_KEY,
+        TimeValue.timeValueMillis(100), TimeValue.timeValueMillis(1), Property.NodeScope);
+
+    public static final Setting<TimeValue> ELECTION_BACK_OFF_TIME_SETTING = Setting.timeSetting(ELECTION_BACK_OFF_TIME_SETTING_KEY,
+        TimeValue.timeValueMillis(100), TimeValue.timeValueMillis(1), Property.NodeScope);
+
+    public static final Setting<TimeValue> ELECTION_MAX_TIMEOUT_SETTING = Setting.timeSetting(ELECTION_MAX_TIMEOUT_SETTING_KEY,
+        TimeValue.timeValueSeconds(10), TimeValue.timeValueMillis(200), Property.NodeScope);
+
+    private final TimeValue initialTimeout;
+    private final TimeValue backoffTime;
+    private final TimeValue maxTimeout;
+    private final ThreadPool threadPool;
+    private final Random random;
+
+    public ElectionSchedulerFactory(Settings settings, Random random, ThreadPool threadPool) {
+        super(settings);
+
+        this.random = random;
+        this.threadPool = threadPool;
+
+        initialTimeout = ELECTION_INITIAL_TIMEOUT_SETTING.get(settings);
+        backoffTime = ELECTION_BACK_OFF_TIME_SETTING.get(settings);
+        maxTimeout = ELECTION_MAX_TIMEOUT_SETTING.get(settings);
+    }
+
+    /**
+     * Start the process to schedule repeated election attempts.
+     *
+     * @param gracePeriod       An initial period to wait before attempting the first election.
+     * @param scheduledRunnable The action to run each time an election should be attempted.
+     */
+    public Releasable startElectionScheduler(TimeValue gracePeriod, Runnable scheduledRunnable) {
+        final ElectionScheduler currentScheduler = new ElectionScheduler();
+        currentScheduler.scheduleNextElection(gracePeriod, scheduledRunnable);
+        return currentScheduler;
+    }
+
+    @SuppressForbidden(reason = "Argument to Math.abs() is definitely not Long.MIN_VALUE")
+    private static long nonNegative(long n) {
+        return n == Long.MIN_VALUE ? 0 : Math.abs(n);
+    }
+
+    /**
+     * @param randomSupplier supplier of randomly-chosen longs
+     * @param upperBound     exclusive upper bound
```

Comment:
also document `@return` here, to denote what's returned? This will go nicely with your explanation above about the uniformly chosen value.

Reply:
😐 it returns a random positive long that's less than `upperBound`.

(actually neater to make it an inclusive upper bound now - this is on it way)